extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQCompactionRunner.java

```diff
@@ -0,0 +1,508 @@
+import org.apache.druid.segment.indexing.DataSchema;
+import org.apache.druid.segment.virtual.ExpressionVirtualColumn;
+import org.apache.druid.sql.calcite.parser.DruidSqlInsert;
+import org.apache.druid.sql.calcite.planner.ColumnMapping;
+import org.apache.druid.sql.calcite.planner.ColumnMappings;
+import org.joda.time.Interval;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class MSQCompactionRunner implements CompactionRunner
+{
+  private static final Logger log = new Logger(MSQCompactionRunner.class);
+  private static final Granularity DEFAULT_SEGMENT_GRANULARITY = Granularities.ALL;
+
+  private final ObjectMapper jsonMapper;
+  private final Injector injector;
+
+  public static final String TIME_VIRTUAL_COLUMN = "__vTime";
+  public static final String TIME_COLUMN = ColumnHolder.TIME_COLUMN_NAME;
+
+  @JsonIgnore
+  private final CurrentSubTaskHolder currentSubTaskHolder = new CurrentSubTaskHolder(
+      (taskObject, config) -> {
+        final MSQControllerTask msqControllerTask = (MSQControllerTask) taskObject;
+        msqControllerTask.stopGracefully(config);
+      });
+
+
+  public MSQCompactionRunner(@JacksonInject ObjectMapper jsonMapper, @JacksonInject Injector injector)
+  {
+    this.jsonMapper = jsonMapper;
+    this.injector = injector;
+  }
+
+  @Override
+  public NonnullPair<Boolean, String> supportsCompactionSpec(
+      CompactionTask compactionTask
+  )
+  {
+    if (compactionTask.getTuningConfig() != null) {
+      PartitionsSpec partitionsSpec = compactionTask.getTuningConfig().getPartitionsSpec();
+      if (!(partitionsSpec instanceof DynamicPartitionsSpec
+            || partitionsSpec instanceof DimensionRangePartitionsSpec)) {
+        return new NonnullPair<>(false, StringUtils.format(
+            "Invalid partition spec type[%s] for MSQ compaction engine."
+            + " Type must be either DynamicPartitionsSpec or DynamicRangePartitionsSpec.",
+            partitionsSpec.getClass()
+        ));
+      }
+      if (partitionsSpec instanceof DynamicPartitionsSpec
+          && ((DynamicPartitionsSpec) partitionsSpec).getMaxTotalRows() != null) {
+        return new NonnullPair<>(false, StringUtils.format(
+            "maxTotalRows[%d] in DynamicPartitionsSpec not supported for MSQ compaction engine.",
+            ((DynamicPartitionsSpec) partitionsSpec).getMaxTotalRows()
+        ));
+      }
+    }
+    if (compactionTask.getMetricsSpec() != null
+        && compactionTask.getGranularitySpec() != null
+        && !compactionTask.getGranularitySpec()
+                          .isRollup()) {
+      return new NonnullPair<>(
+          false,
+          "rollup in granularitySpec must be set to True if metricsSpec is specifed "
+          + "for MSQ compaction engine."
+      );
```

Comment:
```suggestion
      return new NonnullPair<>(
          false,
          "rollup in granularitySpec must be set to true if metricsSpec is specifed "
          + "for MSQ compaction engine."
      );
```

I thought that finalizeAggregations = true, corresponds to rollup = false. Is that not the case?

Reply:
When running with native engine, `rollup=false` doesn't do any aggregation, so the number of rows are as is. It is only when `rollup=true` that the rows are grouped, but in a way equivalent to `finalizeAggregations=false`, leading to metrics being represented with their intermediate types. I believe there is no way to specify `finalizeAggregations=true` equivalent there which is to group but with the final field type for metrics. 
With MSQ, we don't support `rollup=false` if metricsSpec is specified since the metricsSpec is then meaningless. With `rollup=true`, we create a group-by query which can have `finalizeAggregations` as either `true` or `false`. Right now only `true` is supported. The support for `false` will come in a follow-up PR.